### ISSUE TITLE
UnresolvedSubstitution when loading HOCON / conf files

### ISSUE THREAD

**Issue by bkahlert** [sksamuel/hoplite#189]
Steps to reproduce:
1. Remove then question mark in hoplite/hoplite-hocon/src/test/resources/hocon.conf
2. Provide `MONGO_HOSTNAME` by any provider but the environment provider.
3. Run hoplite/hoplite-hocon/src/test/kotlin/com/sksamuel/hoplite/hocon/HoconEnvOverrideTest.kt

```
Run exception com.typesafe.config.ConfigException$UnresolvedSubstitution: Reader: 3: Could not resolve substitution to a value: ${MONGO_HOSTNAME}
```

The reason why the original test passes is because the `MONGO_HOSTNAME` is marked optional. That's why HOCON doesn't throw as it would since it cannot perform any substitutions but by env since that's the only source it uses by default.

It would be great if the options of the parser could be configured.
It would be even if all property providers would load their sources starting with the first and providing that result to the hocon parser when parsing and substituting the second source so the first node already serves as input for the second parsing step. Fold so to say but by providing the intermediary results to the hocon parser.

I used this snippet right before I attempted to move to `hoplite`:
```kotlin
ConfigFactory.systemProperties()
            .withFallback(config)
            .run { fallbacks.fold(this, Config::withFallback) }
            .resolve()
            .extract<MyType>("tld-node")
            .convert()
```
with `config` being an instance of `import com.typesafe.config.Config`, that is the actual config I wanted to load.

**Comment by sksamuel** [comment]
I guess it would be hard to pass data to hocon since the data could be in any format, and we'd have to convert to hocon format, pass in, and then get back out again.

We could make it so that we substitute from source a into source b though, using some hoplite syntax, like {{foo.bar}} ?

**Comment by bkahlert** [comment]
Actually I cannot tell. I read everywhere how great HOCON is but at least when I tried the first time to support that format using the TypeSafe wrapper config4k I experienced that it's not sufficient to provide all possible data sources in a chain of fallbacks. Instead it took me hours to work out the above shown "workaround". I don't even know if that's the supposed solution or indeed just a workaround. That's why I'm not sure I can give much of an advice.
tbh I was hoping to find a more predictable API in hoplite.
(ps: I really like that there is built-in post-processing capabilities in hoplite—something completely missing in type safe)
